### PR TITLE
Autogenerate function descriptions

### PR DESCRIPTION
**Pull request by LinneaAndersson**
In this PR we have created a function to autogenerate the functions tables where we list all functions by category. (this page: https://neo4j.com/docs/cypher-manual/current/functions/). During the build the `GenerateFunctionDescriptions ` creates a document `functionDescription-descriptions-complete.adoc` that contains the tables. The tables are then copied to the file `cypher/cypher-docs/src/docs/dev/ql/functions/index.asciidoc`.

There has been a few changes and I would like your input on whether or not these changes are okay. 

#### Before:
Screenshot 2021-02-03 at 09 57 38

#### After:
Screenshot 2021-02-03 at 09 58 05

#### Before
We created multiple rows in the list for the same function but with different input, as for an example the `date` function. Each row links to one example query. 
 
Screenshot 2021-02-03 at 11 25 47

 
#### After
One row per function - linking to the first example of the function.
Screenshot 2021-02-03 at 11 26 02

### 3) Functions are listed in order

In most cases the function tables are in the same order in this PR and in the current 4.3. But this is not the case in:

- LOAD CSV functions (`file()` and `linenumber()` )
- Temporal functions - duration (`inMonth()`, `inDays()`, `inSeconds()`), is this ok?

### CODE REVIEW

File: cypher/cypher-docs/src/docs/dev/glossary.asciidoc [hunk 1]
```diff
@@ -206,7 +206,7 @@ This section comprises a glossary of all the keywords -- grouped by category and
 |<<functions-pi, pi()>>                         | Trigonometric     | Returns the mathematical constant _pi_.
 |<<functions-point-cartesian-2d, point() - Cartesian 2D>>  | Spatial           | Returns a 2D point object, given two coordinate values in the Cartesian coordinate system.
 |<<functions-point-cartesian-3d,point() - Cartesian 3D>> | Spatial           | Returns a 3D point object, given three coordinate values in the Cartesian coordinate system.
-|<<functions-point-wgs84-2d, point() - WGS 84 2D>>          | Spatial           | Returns a 2D point object, given two coordinate values in the WGS 84 coordinate system.
+|<<functions-point, point() - WGS 84 2D>>          | Spatial           | Returns a 2D point object, given two coordinate values in the WGS 84 coordinate system.
```
Comment (martin-neotech) [inline cypher/cypher-docs/src/docs/dev/glossary.asciidoc]: I would use the `crs` to clean up the `point()` function links.

https://neo4j.com/docs/cypher-manual/current/functions/spatial/#functions-point-cartesian-2d for example have the `csr` `cartesian`

So to cleanup the point() function links would result in:

+ `functions-point-cartesian`
+ `functions-point-cartesian-3d`
+ `functions-point-wgs-84`
+ `functions-point-wgs-84-3d`

And probably do the point function links cleanup in a seperate PR after this one.

File: cypher/cypher-docs/src/docs/dev/ql/functions/index.asciidoc [hunk 2]
```diff
@@ -28,74 +28,79 @@ Related information may be found in <<query-operators>>.
...
 [options="header"]
 |===
-|Function                         |   Description
-|<<functions-all,all()>>          | Tests whether the predicate holds for all elements in a list.
-|<<functions-any,any()>>          | Tests whether the predicate holds for at least one element in a list.
-|<<functions-exists,exists()>>    | Returns true if a match for the pattern exists in the graph, or if the specified property exists in the node, relationship or map.
-|<<functions-isEmpty, isEmpty()>> | Returns true if the given list or map contains no elements or if the given string contains no characters.
-|<<functions-none,none()>>        | Returns true if the predicate holds for no element in a list.
-|<<functions-single,single()>>    | Returns true if the predicate holds for exactly one of the elements in a list.
+| Function | Signature | Description
```
Comment (martin-neotech) [inline cypher/cypher-docs/src/docs/dev/ql/functions/index.asciidoc:38]: Is it possible to use this asciidoc layout style instead?


```
[options="header"]
|===
| Function | Signature | Description

| <<functions-all,`all()`>>
| `all(variable :: VARIABLE IN list :: LIST OF ANY? WHERE predicate :: ANY?) :: (BOOLEAN?)`
| Returns true if the predicate holds for all elements in the given list.

1.3+| <<functions-isempty,`isEmpty()`>>
| `isEmpty(input :: LIST? OF ANY?) :: (BOOLEAN?)`
| Checks whether a list is empty.

//isEmpty()
| `isEmpty(input :: MAP?) :: (BOOLEAN?)`
| Checks whether a map is empty.

//isEmtpy()
| `isEmpty(input :: STRING?) :: (BOOLEAN?)`
| Checks whether a string is empty.

|===
```

**Review by martin-neotech — COMMENTED** [review]
This is great! Only some minor comments.

File: functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            return format( "| %s | %s%n", signature, description );
+        }
+
+        public String firstRow( int sameNameCount )
+        {
+            String referenceName = "functions-" + name.toLowerCase().replace( "()", "" ).replace( '.', '-' );
+            if (referenceExceptions.containsKey( referenceName )) {
+                referenceName = referenceExceptions.get( referenceName );
+            }
+            return format( "1.%s+| <<%s,%s>>  | %s | %s%n",
```
Comment (martin-neotech) [inline functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java]: To make it code styled

```
format( "1.%s+| <<%s,`%s`>>  | `%s` | %s%n",
```

File: functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        public int hashCode()
+        {
+            int result = name.hashCode();
+            result = 31 * result + signature.hashCode();
+            return result;
+        }
+
+        public String row()
+        {
+            return "| <<functions-" + name.toLowerCase().replace( "()", "" ) + "," + name + ">>  | " + description + "\n";
```
Comment (pontusmelke) [inline functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java]: I am not sure if this is a problem but if running on windows the `\n` might cause issues, maybe use `System.lineBreak()` instead.

File: functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+import java.util.Map;
+
+import org.neo4j.dbms.api.DatabaseManagementService;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Result;
+import org.neo4j.graphdb.Transaction;
+
+import static org.neo4j.configuration.GraphDatabaseSettings.DEFAULT_DATABASE_NAME;
+
+public class FunctionDescriptionsGenerator
```
Comment (pontusmelke) [inline functions/src/main/java/org/neo4j/doc/FunctionDescriptionsGenerator.java:42]: This file uses `\n` which might cause rendering issues on windows. Maybe use `System.lineBreak()` or `String.format("...%n")`

**Review by pontusmelke — APPROVED** [review]
Looks good to me. 

My asciidoc is a bit rusty so I didn't look too closely at those changes.

File: cypher/cypher-docs/src/docs/dev/glossary.asciidoc (hunk 1 above)
Comment (LinneaAndersson) [reply cypher/cypher-docs/src/docs/dev/glossary.asciidoc]: Do you mean that you would like to split the `point` function into 4 different rows in the functions table, one for each crs? (like it was before)

Comment (martin-neotech) [reply cypher/cypher-docs/src/docs/dev/glossary.asciidoc]: Yes please.

Comment (LinneaAndersson) [reply cypher/cypher-docs/src/docs/dev/glossary.asciidoc]: I reverted the point functions in the table. The problems I can see with this is that: 
1) We need to do this manually everytime we update the function tables. 
2) It is maybe a bit weird to have the same function (with same signature) on multiple rows in the table.

What do you think? I can leave it like this if you think it's okay :)

**Review by martin-neotech — APPROVED** [review]
Nice!

**Pull request merged**